components/scream/src/share/grid/remap/horizontal_remap_utility.hpp

```diff
@@ -0,0 +1,564 @@
+      auto weights_h     = Kokkos::create_mirror_view(new_seg.weights);
+      Int nsize = seg_match.m_length; 
+      Kokkos::parallel_for("", nsize, KOKKOS_LAMBDA (const int& ii) {
+        new_seg.source_dofs(ii) = seg_match.source_dofs(ii);
+        new_seg.weights(ii)     = seg_match.weights(ii);
+      });
+      Kokkos::fence();
+      nsize = segment.m_length;
+      Kokkos::parallel_for("", nsize, KOKKOS_LAMBDA (const int& ii) {
+        new_seg.source_dofs(seg_match.m_length+ii) = segment.source_dofs(ii);
+        new_seg.weights(seg_match.m_length+ii)     = segment.weights(ii);
+      });
+      Kokkos::fence();
+      map_segments[match_loc] = new_seg;
+    } else {
+      map_segments.push_back(segment);
+    }
+  }; // end add_segment_impl
+/*---------------------------------------------*/
+// This chunk of code should not be needed.  But
+// running into some weird issue with SPA, TODO:
+// Fix this and remove the code below.
+void apply_remap(
+  const view_1d<Real>& source_data,
+        view_1d<Real>& remapped_data)
+{
+  apply_remap<Real>(source_data,remapped_data); 
+}
+/*---------------------------------------------*/
+template <typename ScalarT>
+void apply_remap(
+  const view_3d<ScalarT>& source_data,
+        view_3d<ScalarT>& remapped_data)
+{
+  if (m_num_dofs==0) { return; } // Nothing to do for this rank
+  auto remap_data_h  = Kokkos::create_mirror_view(remapped_data);
+  for (int iseg=0; iseg<get_num_of_segs(); iseg++) {
+    auto seg = map_segments[iseg];
+    auto remap_sub = ekat::subview(remap_data_h,iseg);
+    auto remap_out = seg.apply_segment<ScalarT>(source_data);
+    Kokkos::deep_copy(remap_sub,remap_out);
+  }
+  Kokkos::deep_copy(remapped_data,remap_data_h);
+}
+/*---------------------------------------------*/
+template <typename ScalarT>
+void apply_remap(
+  const view_2d<ScalarT>& source_data,
+        view_2d<ScalarT>& remapped_data)
+{
+  if (m_num_dofs==0) { return; } // Nothing to do for this rank
+  auto remap_data_h  = Kokkos::create_mirror_view(remapped_data);
+  for (int iseg=0; iseg<get_num_of_segs(); iseg++) {
+    auto seg = map_segments[iseg];
+    auto remap_sub = ekat::subview(remap_data_h,iseg);
+    auto remap_out = seg.apply_segment<ScalarT>(source_data);
+    Kokkos::deep_copy(remap_sub,remap_out);
+  }
+  Kokkos::deep_copy(remapped_data,remap_data_h);
+}
+/*---------------------------------------------*/
+template <typename ScalarT>
+void apply_remap(
+  const view_1d<ScalarT>& source_data,
+        view_1d<ScalarT>& remapped_data)
+{
+  if (m_num_dofs==0) { return; } // Nothing to do for this rank
+  auto remap_data_h  = Kokkos::create_mirror_view(remapped_data);
+  for (int iseg=0; iseg<get_num_of_segs(); iseg++) {
+    auto seg = map_segments[iseg];
+    ScalarT remap_val = seg.apply_segment<ScalarT>(source_data);
+    remap_data_h(iseg) = remap_val;
+  }
+  Kokkos::deep_copy(remapped_data,remap_data_h);
+}
+/*---------------------------------------------*/
+
+protected:
```

Comment:
More C++ trivia: in classes that don't belong to an inheritance hierarchy (like this one), `protected` is the same as `private`, so it might be less confusing to use the latter.